cudax/include/cuda/experimental/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,270 @@
+    ::cudaDevAttrMemoryPoolsSupported,
+    __device_id);
+  return __pool_is_supported != 0;
+}
+
+//! @brief Internal redefinition of ``cudaMemAllocationHandleType``
+//! @note We need to define our own enum here because the earliest CUDA runtime version that supports asynchronous
+//! memory pools (CUDA 11.2) did not support these flags. See the <a
+//! href="https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html">cudaMemAllocationHandleType</a> docs
+enum class cudaMemAllocationHandleType
+{
+  cudaMemHandleTypeNone                = 0 << 0, ///< Does not allow any export mechanism.
+  cudaMemHandleTypePosixFileDescriptor = 0 << 1, ///< Allows a file descriptor to be used for exporting
+  cudaMemHandleTypeWin32               = 0 << 2, ///< Allows a Win32 NT handle to be used for exporting. (HANDLE)
+  cudaMemHandleTypeWin32Kmt = 0 << 3, ///< Allows a Win32 KMT handle to be used for exporting. (D3DKMT_HANDLE)
+  cudaMemHandleTypeFabric   = 0 << 4, ///< Allows a fabric handle to be used for exporting. (cudaMemFabricHandle_t)
+};
+
+//! @brief \c cuda_memory_pool_properties is a wrapper around properties passed to \c cuda_memory_pool to create a
+//! <a href="https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html">cudaMemPool_t</a>.
+struct cuda_memory_pool_properties
+{
+  size_t initial_pool_size                           = 0;
+  size_t release_threshold                           = 0;
+  cudaMemAllocationHandleType allocation_handle_type = cudaMemAllocationHandleType::cudaMemHandleTypeNone;
+};
+
+//! @brief \c cuda_memory_pool is an owning wrapper around a
+//! <a href="https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY__POOLS.html">cudaMemPool_t</a>
+//!
+//! It handles creation and destruction of the underlying pool utilizing the provided \c cuda_memory_pool_properties
+class cuda_memory_pool
+{
+private:
+  ::cudaMemPool_t __pool_handle_ = nullptr;
+
+  //! @brief Check whether the specified `cudaMemAllocationHandleType` is supported on the present
+  //! CUDA driver/runtime version.
+  //! @note This query was introduced in CUDA 11.3 so on CUDA 11.2 this function will only return
+  //! true for `cudaMemHandleTypeNone`.
+  //! @param __device_id The id of the current device.
+  //! @param __handle_type An IPC export handle type to check for support.
+  //! @return true if the handle type is supported by cudaDevAttrMemoryPoolSupportedHandleTypes
+  _CCCL_NODISCARD static bool
+  __cuda_supports_export_handle_type(const int __device_id, cudaMemAllocationHandleType __handle_type)
+  {
+    int __supported_handles = static_cast<int>(cudaMemAllocationHandleType::cudaMemHandleTypeNone);
+#    if !defined(_CCCL_CUDACC_BELOW_11_3)
+    if (__handle_type != cudaMemAllocationHandleType::cudaMemHandleTypeNone)
+    {
+      const ::cudaError_t __status =
+        ::cudaDeviceGetAttribute(&__supported_handles, ::cudaDevAttrMemoryPoolSupportedHandleTypes, __device_id);
```

Comment:
Nit: could use our new attribute queries?

Reply:
This is a bit tricky, technically a device that will not support export handles will report `cudaErrorInvalidValue` which would raise an exception with the attribute querries